File: backend/app/schemas/calls.py
"""Pydantic schemas for call log and artifact endpoints."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, field_validator


class CallEventResponse(BaseModel):
    id: str
    event_type: str
    provider_status: str | None = None
    event_at: datetime

    model_config = {"from_attributes": True}


class CallListItem(BaseModel):
    id: str
    created_at: datetime
    direction: str
    from_masked: str
    to_masked: str
    status: str
    duration_seconds: int | None = None
    source_type: str
    missing_summary: bool
    missing_transcript: bool
    missing_labels: bool
    started_at: datetime
    artifact_status: str | None = None
    agent_id: str | None = None
    voice_id: str | None = None
    booked_calendar_event_id: str | None = None
    booked_calendar_event_summary: str | None = None
    caller_display_name: str | None = None
    caller_relationship: str | None = None
    is_vip: bool = False
    is_blocked: bool = False
    has_notes: bool = False
    has_reminder: bool = False
    labels: list[str] = []

    model_config = {"from_attributes": True}


class CallListResponse(BaseModel):
    items: list[CallListItem]
    next_cursor: str | None = None
    has_more: bool = False


class CallDetailResponse(BaseModel):
    id: str
    direction: str
    source_type: str
    from_masked: str
    to_masked: str
    status: str
    started_at: datetime
    ended_at: datetime | None = None
    duration_seconds: int | None = None
    forwarding_detected: bool
    missing_summary: bool
    missing_transcript: bool
    missing_labels: bool
    events: list[CallEventResponse] = []
    created_at: datetime
    summary: str | None = None
    summary_status: str | None = None
    labels: list[LabelResponse] | None = None
    labels_status: str | None = None
    transcript_status: str | None = None
    notes: str | None = None
    recording_available: bool = False
    is_vip: bool = False
    is_blocked: bool = False
    caller_display_name: str | None = None
    caller_relationship: str | None = None
    agent_id: str | None = None
    voice_id: str | None = None
    booked_calendar_event_id: str | None = None
    booked_calendar_event_summary: str | None = None
    booked_appointment_date: str | None = None
    booked_appointment_time: str | None = None
    booked_appointment_duration_minutes: int | None = None
    booked_appointment_caller_name: str | None = None
    booked_appointment_reason: str | None = None

    model_config = {"from_attributes": True}


class MarkBlockedRequest(BaseModel):
    reason: str | None = None


class MarkStatusResponse(BaseModel):
    is_vip: bool = False
    is_blocked: bool = False


class LabelResponse(BaseModel):
    label_name: str
    reason_text: str
    evidence_snippets: list[str] = []
    confidence: float = 0.0
    spam_score: float | None = None
    produced_by: str = "deterministic"


class ArtifactResponse(BaseModel):
    call_id: str
    summary: str | None = None
    summary_status: str
    labels: list[LabelResponse] = []
    labels_status: str
    transcript_status: str
    structured_extraction: dict | None = None


class TranscriptTurn(BaseModel):
    role: str
    text: str
    time_seconds: float = 0.0


class TranscriptResponse(BaseModel):
    call_id: str
    conversation_id: str | None = None
    turns: list[TranscriptTurn] = []
    turn_count: int = 0
    status: str = "ready"


_URGENCY_VALUES = {"low", "normal", "urgent"}
_SPAM_LABEL_VALUES = {"normal_call", "possible_spam", "sales", "unknown"}


class CallPatchRequest(BaseModel):
    urgency_level: str | None = None
    spam_label: str | None = None
    is_vip_override: bool | None = None
    notes: str | None = None

    @field_validator("urgency_level")
    @classmethod
    def validate_urgency(cls, v: str | None) -> str | None:
        if v is not None and v not in _URGENCY_VALUES:
            raise ValueError(f"urgency_level must be one of {sorted(_URGENCY_VALUES)}")
        return v

    @field_validator("spam_label")
    @classmethod
    def validate_spam_label(cls, v: str | None) -> str | None:
        if v is not None and v not in _SPAM_LABEL_VALUES:
            raise ValueError(f"spam_label must be one of {sorted(_SPAM_LABEL_VALUES)}")
        return v

    @field_validator("notes")
    @classmethod
    def validate_notes_length(cls, v: str | None) -> str | None:
        if v is not None and len(v) > 5000:
            raise ValueError("notes must be at most 5000 characters")
        return v


class SummaryRegenerateResponse(BaseModel):
    call_id: str
    summary_status: str
    message: str


class RecordingResponse(BaseModel):
    call_id: str
    recording_available: bool = False
    recording_url: str | None = None
    duration_seconds: int | None = None


class MemoryItemResponse(BaseModel):
    id: str
    memory_type: str
    subject: str | None = None
    value: str | None = None
    confidence: float | None = None
    user_confirmed: bool = False
    source_call_id: str | None = None
    caller_phone_hash: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MemoryListResponse(BaseModel):
    items: list[MemoryItemResponse]


_ALLOWED_MEMORY_TYPES = {
    "callback_window_preference",
    "caller_display_name",
    "communication_preference",
    "relationship_tag",
    "repeated_reason_pattern",
}


class CreateMemoryItemRequest(BaseModel):
    memory_type: str
    subject: str
    value: str
    caller_phone_hash: str | None = None

    @field_validator("memory_type")
    @classmethod
    def validate_memory_type(cls, v: str) -> str:
        if v not in _ALLOWED_MEMORY_TYPES:
            raise ValueError(f"memory_type must be one of {sorted(_ALLOWED_MEMORY_TYPES)}")
        return v

    @field_validator("subject")
    @classmethod
    def validate_subject(cls, v: str) -> str:
        if len(v) > 100:
            raise ValueError("subject must be at most 100 characters")
        return v

    @field_validator("value")
    @classmethod
    def validate_value(cls, v: str) -> str:
        if len(v) > 2000:
            raise ValueError("value must be at most 2000 characters")
        return v


class UpdateMemoryItemRequest(BaseModel):
    subject: str | None = None
    value: str | None = None
    user_confirmed: bool | None = None

    @field_validator("subject")
    @classmethod
    def validate_subject(cls, v: str | None) -> str | None:
        if v is not None and len(v) > 100:
            raise ValueError("subject must be at most 100 characters")
        return v

    @field_validator("value")
    @classmethod
    def validate_value(cls, v: str | None) -> str | None:
        if v is not None and len(v) > 2000:
            raise ValueError("value must be at most 2000 characters")
        return v

File: mobile/src/api/calls.ts
import { apiClient } from './client';

export interface CallEvent {
  id: string;
  event_type: string;
  provider_status: string | null;
  event_at: string;
}

export interface CallLabel {
  label_name: string;
  reason_text: string;
  evidence_snippets: string[];
  confidence: number;
  produced_by: string;
}

export interface CallListItem {
  id: string;
  created_at: string;
  direction: string;
  from_masked: string;
  to_masked: string;
  status: string;
  duration_seconds: number | null;
  source_type: string;
  missing_summary: boolean;
  missing_transcript: boolean;
  missing_labels: boolean;
  started_at: string;
  artifact_status: string | null;
  caller_display_name?: string | null;
  caller_relationship?: string | null;
  is_vip?: boolean;
  is_blocked?: boolean;
  booked_calendar_event_id?: string | null;
  booked_calendar_event_summary?: string | null;
}

export interface CallListResponse {
  items: CallListItem[];
  next_cursor: string | null;
  has_more: boolean;
}

export interface CallDetail {
  id: string;
  direction: string;
  source_type: string;
  from_masked: string;
  to_masked: string;
  status: string;
  started_at: string;
  ended_at: string | null;
  duration_seconds: number | null;
  forwarding_detected: boolean;
  missing_summary: boolean;
  missing_transcript: boolean;
  missing_labels: boolean;
  events: CallEvent[];
  created_at: string;
  summary: string | null;
  summary_status: string | null;
  labels: CallLabel[] | null;
  labels_status: string | null;
  transcript_status: string | null;
  notes: string | null;
  recording_available?: boolean;
}

export interface CallArtifacts {
  call_id: string;
  summary: string | null;
  summary_status: string;
  labels: CallLabel[];
  labels_status: string;
  transcript_status: string;
  structured_extraction: Record<string, unknown> | null;
}

export interface TranscriptTurn {
  role: string;
  text: string;
  time_seconds: number;
}

export interface TranscriptResponse {
  call_id: string;
  conversation_id: string | null;
  turns: TranscriptTurn[];
  turn_count: number;
  status: string;
}

export interface CallFilters {
  label?: string;
  search?: string;
  status?: string;
  source_type?: string;
  date_from?: string;
  date_to?: string;
  duration_min?: number;
  duration_max?: number;
  country_prefix?: string;
  has_recording?: boolean;
  sort_by?: string;
  sort_dir?: string;
}

export async function fetchCalls(cursor?: string, filters?: CallFilters): Promise<CallListResponse> {
  const params: Record<string, string> = {};
  if (cursor) params.cursor = cursor;
  if (filters?.label) params.label = filters.label;
  if (filters?.search) params.search = filters.search;
  if (filters?.status) params.status = filters.status;
  if (filters?.source_type) params.source_type = filters.source_type;
  if (filters?.date_from) params.date_from = filters.date_from;
  if (filters?.date_to) params.date_to = filters.date_to;
  if (filters?.duration_min != null) params.duration_min = String(filters.duration_min);
  if (filters?.duration_max != null) params.duration_max = String(filters.duration_max);
  if (filters?.country_prefix) params.country_prefix = filters.country_prefix;
  if (filters?.has_recording === true) params.has_recording = 'true';
  if (filters?.sort_by) params.sort_by = filters.sort_by;
  if (filters?.sort_dir) params.sort_dir = filters.sort_dir;
  const { data } = await apiClient.get('/calls', { params });
  return data;
}

export async function fetchCallerPhone(callId: string): Promise<{ phone: string }> {
  const { data } = await apiClient.get(`/calls/${callId}/caller-phone`);
  return data;
}

export async function fetchCallDetail(callId: string): Promise<CallDetail> {
  const { data } = await apiClient.get(`/calls/${callId}`);
  return data;
}

export async function fetchCallArtifacts(callId: string): Promise<CallArtifacts> {
  const { data } = await apiClient.get(`/calls/${callId}/artifacts`);
  return data;
}

export async function fetchCallTranscript(callId: string): Promise<TranscriptResponse> {
  const { data } = await apiClient.get(`/calls/${callId}/transcript`);
  return data;
}

export async function retryCallTranscript(callId: string): Promise<{ status: string; message: string }> {
  const { data } = await apiClient.post(`/calls/${callId}/transcript/retry`);
  return data;
}

export async function patchCallNotes(callId: string, notes: string): Promise<CallDetail> {
  const { data } = await apiClient.patch(`/calls/${callId}`, { notes });
  return data;
}

export function getCallRecordingUrl(callId: string, token: string): string {
  const base = apiClient.defaults.baseURL ?? '';
  return `${base}/calls/${callId}/recording?token=${encodeURIComponent(token)}`;
}

export async function markCallVip(callId: string): Promise<void> {
  await apiClient.post(`/calls/${callId}/mark-vip`);
}

export async function unmarkCallVip(callId: string): Promise<void> {
  await apiClient.delete(`/calls/${callId}/mark-vip`);
}

export async function markCallBlocked(callId: string, reason?: string): Promise<void> {
  await apiClient.post(`/calls/${callId}/mark-blocked`, reason ? { reason } : {});
}

export async function unmarkCallBlocked(callId: string): Promise<void> {
  await apiClient.delete(`/calls/${callId}/mark-blocked`);
}
